### ISSUE TITLE
Undoing move finished generation without my consent in solo mode

### ISSUE THREAD

**Issue by trzycztery** [terraforming-mars/terraforming-mars#2630]
I've been doing some moves in gen 10, decided to undo two moves, without refreshing page. After second undo I saw World Terraforming menu, even though I didn't pass. Then I was taken to gen 11 like I've had passed.
https://terraforming-mars.herokuapp.com/player?id=718996314061

**Comment by kevinb9n** [comment]
Yeah this one hurts.

**Comment by Lynesth** [comment]
> decided to undo two moves, without refreshing page

Something tells me you knew you should have refreshed.
Here's what happened.

Let's say the actions were like this:
- Pass
- Sell Patents
- Undo

And you selected **Undo**.

Then you were presented with the exact same list of actions and you selected **Undo** again.
Well, that's what you _think_ you selected.
But if you instead refreshed the page, you would have noticed that the menu was actually:
- Play Project Card
- Standard Project
- Pass
- Sell Patents
- Undo

And you selected the 3rd item in the list (that's how the server knows what you selected), which is... **Pass**.

I'll try work on an auto refresh this week if I have time. Meanwhile, the rule is:

**Undid** an action ? **Refresh** the page.

**Comment by kevinb9n** [comment]
Count me also as "well aware of the rule, but occasionally messes up" - and it's just that the consequence in this case is so severe. Oh, and if there isn't a quick fix, maybe the text could say "Undo last action (NOTE: you must reload the page next!)" for now? Because it's hard to get this message out to everyone otherwise.

**Comment by bafolts** [comment]
We don't need to necessarily refresh the page. After processing the input for the undo action we need to return the latest player object. There is a moderately fast fix here I can put in.

**Comment by kberg** [comment]
@bafolts since you know what to do here, can you knock it out?

**Comment by Borbarad2** [comment]
What is left to do here? Is the bug fixed or not?

**Comment by kberg** [comment]
It is not fixed yet.

**Comment by kevinb9n** [comment]
Huh, really? I haven't seen this in a very long time.

**Comment by kberg** [comment]
That's great. I haven't dug into it, though.

**Comment by katieyang** [comment]
I think this bug is fixed. Couldn't recreate it with 1 player World Terraforming + Allow Undo.

**Comment by kberg** [comment]
Yeah, I'll just close it. I won't look into it right now anyway. Thank you, everyone.
